ROX-23861: Add close delimiter for Email MIME boundary

Skalador opened the pull request:
## Description

Added closing delimiter `--` for the last iteration in `email.go` for Email MIME. Updated the unit tests in `email_test.go` to verify the final closing boundary has the following structure: `--boundaryValue--`.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Modified the unit test `TestEmailMsgWithAttachment` in `central/notifiers/email/email_test.go` which checks for the close delimiter via `assert.Equal(t, expectedFinalBoundary, lastBoundary)`. The unit tests passes: 
```sh
go test -v ./central/notifiers/email
=== RUN   TestEmailMsgWithAttachment
--- PASS: TestEmailMsgWithAttachment (0.00s)
```

Note: It probably would make sense to add a unit test with multiple attachments to verify the final closing boundary (worked on my temporary tests by modifying the `msg` in `TestEmailMsgWithAttachment`).

charmik-redhat commented on central/notifiers/email/email.go:
```diff
@@ -256,6 +256,8 @@ func (m Message) writeContentBytes(buf *bytes.Buffer) {
 	writer := multipart.NewWriter(buf)
 	boundary := writer.Boundary()
+	var idx int
+	numberOfAttachments := len(m.Attachments)
```
I think we can have `multipart/mixed` content type even if there are 0 attachments. So, instead of `numAttachments`, you can use a `bool mixedType` like below
```
var bool mixedType
if m.EmbedLogo || len(m.Attachments) > 0 {
    mixedType = true
}
if mixedType {
    buf.WriteString(fmt.Sprintf("Content-Type: multipart/mixed; boundary=\"%s\"\r\n", boundary))
    buf.WriteString(fmt.Sprintf("\n--%s\r\n", boundary))
}
```

Skalador replied:
@charmik-redhat Thanks for reviewing. I like your suggestion and will update the code accordingly. Additionally, I will add test cases with multiple attachments and no attachments, but with logo. Consequently we will have more test coverage.

charmik-redhat commented on central/notifiers/email/email.go:
```diff
@@ -256,6 +256,8 @@ func (m Message) writeContentBytes(buf *bytes.Buffer) {
+	numberOfAttachments := len(m.Attachments)
 
 	if m.EmbedLogo {
 		buf.WriteString(fmt.Sprintf("Content-Type: multipart/mixed; boundary=\"%s\"\r\n", boundary))
```
After setting content type and adding boundary using `bool mixedType` as per previous comment, you can remove this and line below.

charmik-redhat commented on central/notifiers/email/email.go:
```diff
@@ -278,12 +280,17 @@ func (m Message) writeContentBytes(buf *bytes.Buffer) {
 		buf.WriteString(fmt.Sprintf("\r\n%s\r\n", applyRfc5322LineLengthLimit(base64.StdEncoding.EncodeToString(v))))
-		buf.WriteString(fmt.Sprintf("\n--%s\r\n", boundary))
+		if idx == numberOfAttachments {
+			buf.WriteString(fmt.Sprintf("\n--%s--\r\n", boundary))
```
With `bool mixedType` , you can move this out of the for loop so that it will add the correct final boundary even when there are 0 attachments.

```
if mixedType {
    buf.WriteString(fmt.Sprintf("\n--%s--\r\n", boundary))
}
```

charmik-redhat reviewed (commented):
Thanks @Skalador for contributing this bug fix to the repository. I suggested some small changes to your fix so that we can avoid the same issue for emails without attachments too. Feel free to reply in comments if you hav any questions.

Skalador commented:
Hi @charmik-redhat, I have added your suggestions. Additionally, I have added unit tests to cover multiple attachments `TestEmailMsgWithMultipleAttachments` and no attachment with logo `TestEmailMsgNoAttachmentsWithLogo`: 
```sh
go test -v ./central/notifiers/email                 
=== RUN   TestEmailMsgWithAttachment
--- PASS: TestEmailMsgWithAttachment (0.00s)
=== RUN   TestEmailMsgWithMultipleAttachments
--- PASS: TestEmailMsgWithMultipleAttachments (0.00s)
=== RUN   TestEmailMsgNoAttachmentsWithLogo
--- PASS: TestEmailMsgNoAttachmentsWithLogo (0.00s)
=== RUN   TestEmailMsgNoAttachments
--- PASS: TestEmailMsgNoAttachments (0.00s)
...
PASS
ok  	github.com/stackrox/rox/central/notifiers/email
```

Lastly, I updated the `CHANGELOG` with `ROX-23861: Add close delimiter for Email MIME boundary.`.

Please take a look and feel free to edit/change.

charmik-redhat commented:
/ok-to-test

charmik-redhat commented on central/notifiers/email/email_test.go:
```diff
@@ -48,6 +51,107 @@ func TestEmailMsgWithAttachment(t *testing.T) {
+	// Obtain boundary and verify the close delimiter
+	regex := regexp.MustCompile(` boundary="([^"]+)"`)
+	// Regex will provide two values [boundary="boundaryValue", boundaryValue]
+	boundary := regex.FindStringSubmatch(msgStr)
```
Use `regex.FindString(msgStr)` . That will match `boundary="boundaryValue"`. Then you can split by '=' and remove double quotes from splitResults[1] to get `boundaryValue`. That will make it more clear to understand what is being selected here.

Skalador replied:
Thanks for this suggestion, this makes the code more clear.

charmik-redhat reviewed (commented):
Few small comments. Otherwise changes look good.

Skalador commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
 - ROX-18689: ACS will qualify the registry (and path) of images from the container runtime when env var `ROX_UNQUALIFIED_SEARCH_REGISTRIES` is set to `true` on both Central and Sensor.
   - This enables support for CRI-O's unqualified search registries and short name aliases ([more info](https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md)).
 - ROX-23852: `roxctl image scan` now has the option to filter by vulnerability severities using the `--severity` flag.
+- ROX-23861: Add close delimiter for Email MIME boundary.
```
okay, I will revert this change.

Skalador commented on central/notifiers/email/email.go:
```diff
@@ -256,8 +256,12 @@ func (m Message) writeContentBytes(buf *bytes.Buffer) {
+	if m.EmbedLogo || len(m.Attachments) > 0 {
+		mixedType = true
+	}
+	if mixedType {
```
Makes sense.

Skalador commented on central/notifiers/email/email.go:
```diff
@@ -256,8 +256,12 @@ func (m Message) writeContentBytes(buf *bytes.Buffer) {
+		mixedType = true
+	}
+	if mixedType {
 		buf.WriteString(fmt.Sprintf("Content-Type: multipart/mixed; boundary=\"%s\"\r\n", boundary))
```
Will do.

Skalador commented on central/notifiers/email/email_test.go:
```diff
@@ -48,6 +51,107 @@ func TestEmailMsgWithAttachment(t *testing.T) {
 	assert.Contains(t, msgStr, "<div>\r\nHow you doin'?\r\n</div>\r\n")
+
+	lastBoundary, expectedFinalBoundary, err := obtainLastAndExpectedBoundaryString(msgStr)
+	if err != nil {
```
That is way better, I agree.

Skalador commented:
Sorry @charmik-redhat , I have tried to rebase this branch to fix the CI issues as suggested. I don't think that's how it is supposed to be.

charmik-redhat commented:
Looks like a rebase messed something up. Can you close this and create a new PR with changes?

Skalador commented:
> Looks like a rebase messed something up. Can you close this and create a new PR with changes?

Sure, just created #11053.
